processing/src/main/java/io/druid/query/groupby/GroupByQuery.java

```diff
@@ -251,46 +274,78 @@ public int compare(Object lhs, Object rhs)
 
   public Ordering<Row> getRowOrdering(final boolean granular)
   {
-    final Comparator naturalNullsFirst = Ordering.natural().nullsFirst();
+    final boolean sortByDimsFirst = getContextSortByDimsFirst();
 
-    return Ordering.from(
-        new Comparator<Row>()
-        {
-          @Override
-          public int compare(Row lhs, Row rhs)
-          {
-            final int timeCompare;
+    final Comparator<Row> timeComparator = getTimeComparator(granular);
 
-            if (granular) {
-              timeCompare = Longs.compare(
-                  granularity.truncate(lhs.getTimestampFromEpoch()),
-                  granularity.truncate(rhs.getTimestampFromEpoch())
-              );
-            } else {
-              timeCompare = Longs.compare(
-                  lhs.getTimestampFromEpoch(),
-                  rhs.getTimestampFromEpoch()
-              );
-            }
+    if (sortByDimsFirst) {
+      return Ordering.from(
+          new Comparator<Row>()
+          {
+            @Override
+            public int compare(Row lhs, Row rhs)
+            {
+              final int cmp = compareDims(NATURAL_NULLS_FIRST, dimensions, lhs, rhs);
+              if (cmp != 0) {
+                return cmp;
+              }
 
-            if (timeCompare != 0) {
-              return timeCompare;
+              return timeComparator.compare(lhs, rhs);
             }
+          }
+      );
+    } else {
+      return Ordering.from(
+          new Comparator<Row>()
+          {
+            @Override
+            public int compare(Row lhs, Row rhs)
+            {
+              final int timeCompare = timeComparator.compare(lhs, rhs);
 
-            for (DimensionSpec dimension : dimensions) {
-              final int dimCompare = naturalNullsFirst.compare(
-                  lhs.getRaw(dimension.getOutputName()),
-                  rhs.getRaw(dimension.getOutputName())
-              );
-              if (dimCompare != 0) {
-                return dimCompare;
+              if (timeCompare != 0) {
+                return timeCompare;
               }
-            }
 
-            return 0;
+              return compareDims(NATURAL_NULLS_FIRST, dimensions, lhs, rhs);
+            }
           }
+      );
+    }
+  }
+
+  private Comparator<Row> getTimeComparator(boolean granular)
+  {
+    if (granular) {
+      return new Comparator<Row>()
+      {
+        @Override
+        public int compare(Row lhs, Row rhs)
+        {
+          return Longs.compare(
+              granularity.truncate(lhs.getTimestampFromEpoch()),
+              granularity.truncate(rhs.getTimestampFromEpoch())
+          );
         }
-    );
+      };
+    } else {
+      return NON_GRANULAR_TIME_COMP;
+    }
+  }
+
+  private static int compareDims(Comparator naturalNullsFirst, List<DimensionSpec> dimensions, Row lhs, Row rhs)
```

Comment:
What is the reason for `naturalNullsFirst` to be a parameter? Why cannot just access static inside the method?